pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/PlanNodeToRelConverter.java

```diff
@@ -0,0 +1,468 @@
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.RelDistribution;
+import org.apache.calcite.rel.RelDistributions;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.SetOp;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rel.logical.LogicalIntersect;
+import org.apache.calcite.rel.logical.LogicalMinus;
+import org.apache.calcite.rel.logical.LogicalSort;
+import org.apache.calcite.rel.logical.LogicalUnion;
+import org.apache.calcite.rel.logical.LogicalWindow;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexWindowBound;
+import org.apache.calcite.rex.RexWindowBounds;
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.pinot.common.utils.DatabaseUtils;
+import org.apache.pinot.core.operator.ExplainAttributeBuilder;
+import org.apache.pinot.core.plan.PinotExplainedRelNode;
+import org.apache.pinot.query.planner.plannode.AggregateNode;
+import org.apache.pinot.query.planner.plannode.ExchangeNode;
+import org.apache.pinot.query.planner.plannode.ExplainedNode;
+import org.apache.pinot.query.planner.plannode.FilterNode;
+import org.apache.pinot.query.planner.plannode.JoinNode;
+import org.apache.pinot.query.planner.plannode.MailboxReceiveNode;
+import org.apache.pinot.query.planner.plannode.MailboxSendNode;
+import org.apache.pinot.query.planner.plannode.PlanNode;
+import org.apache.pinot.query.planner.plannode.PlanNodeVisitor;
+import org.apache.pinot.query.planner.plannode.ProjectNode;
+import org.apache.pinot.query.planner.plannode.SetOpNode;
+import org.apache.pinot.query.planner.plannode.SortNode;
+import org.apache.pinot.query.planner.plannode.TableScanNode;
+import org.apache.pinot.query.planner.plannode.ValueNode;
+import org.apache.pinot.query.planner.plannode.WindowNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Converts a {@link PlanNode} into a {@link RelNode}.
+ *
+ * This class is used to convert serialized plan nodes into RelNodes so they can be used when explain with
+ * implementation is requested. Therefore some nodes may be transformed in a way that loses information that is
+ * required to create an actual executable plan but not necessary in order to describe the plan.
+ */
+public final class PlanNodeToRelConverter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PlanNodeToRelConverter.class);
+
+  private PlanNodeToRelConverter() {
+  }
+
+  public static RelNode convert(RelBuilder builder, PlanNode planNode) {
+    ConverterVisitor visitor = new ConverterVisitor(builder);
+    planNode.visit(visitor, null);
+
+    return visitor.build();
+  }
+
+  private static class ConverterVisitor implements PlanNodeVisitor<Void, Void> {
```

Comment:
It is similar to the other visitor I've converted. The fact that the implementation uses a visitor is an implementation detail we don't need to leak into the caller. Imagine the case you mention. Would you be able to know if you can reuse the same visitor in two different calls? Why would you need to call builder.build() in your code? What is easier to read?

```java
Builder builder = ...;
PlanNode planNode = ...;
RelNode relNode = PlanNodeToRelConverter.convert(builder, planNode);
```

or

```java
Builder builder = ...;
PlanNode planNode = ...;
PlanNodeToRelConverter visitor = new PlanNodeToRelConverter.convert(builder, planNode);
planNode.visit(visitor, null);
RelNode relNode = builder.build();
```

Also... the `convert` method has no javadoc but in case we have to add it, it is simpler to add it there.

I love visitors (although I would love even more to be able to use sealed classes) but they are a poor man switch. Switches are great tools to use in your functions, but in Java we offer methods to be called, not switches.

Reply:
Makes sense, thanks for elaborating! 😄